plugins/modules/foreman_snapshot.py

```diff
@@ -0,0 +1,155 @@
+
+DOCUMENTATION = '''
+---
+module: foreman_snapshot
+short_description: Manage Foreman Snapshots
+description:
+  - "Manage Foreman Snapshots for Host Entities"
+  - "This beta version can create, update, revert and delete snapshots from preexisting host"
+author:
+  - "Manisha Singhal (@Manisha15) ATIX AG"
+requirements:
+    - "apypie >= 0.0.1"
+options:
+  server_url:
+    description:
+      - URL of Foreman server
+    required: true
+  username:
+    description:
+      - Username on Foreman server
+    required: true
+  password:
+    description:
+      - Password for user accessing Foreman server
+    required: true
+  validate_certs:
+    aliases: [ verify_ssl ]
+    description:
+      - Verify SSL of the Foreman server
+    required: false
+    default: true
+    type: bool
+  name:
+    description:
+      - Name of Snapshot
+    required: true
+  description:
+    description:
+      - Description of Snapshot
+    required: false
+  host:
+    description:
+      - Name of related Host
+    required: true
+
+'''
+
+EXAMPLES = '''
+- name: "Create a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: present
+
+- name: "Update a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    description: "description"
+    state: present
+
+- name: "Revert a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: reverted
+
+- name: "Delete a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: absent
+'''
+
+RETURN = ''' # '''
+
+
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule
+
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'description',
+    'host_id': 'host_id',
+}
+
+
+def main():
+    module = ForemanEntityApypieAnsibleModule(
+        argument_spec=dict(
+            name=dict(required=True),
+            description=dict(),
+            host=dict(required=True),
+            host_id=dict(),
+            state=dict(default='present', choices=['present', 'absent', 'reverted']),
+        ),
+        supports_check_mode=True,
+    )
+    (snapshot_dict, state) = module.parse_params()
+
+    module.connect()
+
+    host = module.find_resource_by_name('hosts', name=snapshot_dict['host'], failsafe=False, thin=True)
+
+    snapshot_dict['host_id'] = host['id']
+
+    entity = module.find_resource_by_name('snapshots', name=snapshot_dict['name'], params={'host_id': host['id']}, failsafe=True)
+
+    if state == 'reverted':
```

Comment:
can this state also be done via `ensure_resource_state`? by passing the special dict you generate maybe?

Reply:
It is not working with ensure_resource_state. Revert action for snapshot is working weirdly. It has required parameters id and host_id but it throws HTTPError for that but if you pass additional parameter with while calling rever, say, name or description, the revert function works. In `ensure_resource_state` it uses resource payload which removes the additional parameters (here name or description) causing HTTPError.

Reply:
aaaaaaaaah ok

I'd argue this is an Apypie bug or an API bug, or whatever. Then your workaround if fine, but I'll think about how we can make it nicer in Apypie.